# Calculate format=diff pkg(sys-kernel/calculate-sources)=>4.6,4.7&&merge(x11-drivers/nvidia-drivers)=>340.0,341
diff -Nuar nvidia-drivers-340.96-r5.old/kernel/os-mlock.c nvidia-drivers-340.96-r5/kernel/os-mlock.c
--- nvidia-drivers-340.96-r5.old/kernel/os-mlock.c      2015-11-09 08:44:53.000000000 +0300
+++ nvidia-drivers-340.96-r5/kernel/os-mlock.c  2016-08-12 14:27:21.113946450 +0300
@@ -45,7 +45,7 @@
     }

     down_read(&mm->mmap_sem);
-    ret = get_user_pages(current, mm, (unsigned long)address,
+    ret = get_user_pages_remote(current, mm, (unsigned long)address,
             page_count, write, force, user_pages, NULL);
     up_read(&mm->mmap_sem);
     pinned = ret;
@@ -58,7 +58,7 @@
     else if (pinned < page_count)
     {
         for (i = 0; i < pinned; i++)
-            page_cache_release(user_pages[i]);
+            put_page(user_pages[i]);
         os_free_mem(user_pages);
         return RM_ERR_INVALID_ADDRESS;
     }
@@ -85,7 +85,7 @@
     {
         if (write)
             set_page_dirty_lock(user_pages[i]);
-        page_cache_release(user_pages[i]);
+        put_page(user_pages[i]);
     }

     os_free_mem(user_pages);
diff -Nuar nvidia-drivers-340.96-r5.old/kernel/uvm/nvidia_uvm_lite.c nvidia-drivers-340.96-r5/kernel/uvm/nvidia_uvm_lite.c
--- nvidia-drivers-340.96-r5.old/kernel/uvm/nvidia_uvm_lite.c   2015-11-09 08:44:53.000000000 +0300
+++ nvidia-drivers-340.96-r5/kernel/uvm/nvidia_uvm_lite.c       2016-08-12 14:27:53.680945901 +0300
@@ -785,7 +785,7 @@
         // a reference so that the fault handling logic is correct:
         //
         get_page(pTracking->uvmPage);
-        retValue = VM_FAULT_MINOR;
+        retValue = 0;
     }

     pRecord->isMapped = NV_TRUE;
